src/content/popup/popup.ts

```diff
@@ -492,3 +498,91 @@ function getPopupArrow(): HTMLElement | null {
     ? hostElem.shadowRoot.querySelector('.arrow')
     : null;
 }
+
+type HorizontalSwipeDirection = 'right' | 'left';
+
+function onHorizontalSwipe(
+  element: HTMLElement,
+  handler: (swipeDirection: HorizontalSwipeDirection) => void
+) {
+  // Min x distance traveled to be considered swipe
+  const xMinThreshold = 150;
+  // Max y distance that can be traveled before it is no longer considered a horizontal swipe
+  const yMaxThreshold = 100;
+  // Max time allowed to travel that distance
+  const allowedTime = 200;
+
+  let startTime = 0;
+  let startX: number;
+  let startY: number;
+
+  element.addEventListener(
+    'touchstart',
+    function (e) {
+      startX = e.changedTouches[0].pageX;
+      startY = e.changedTouches[0].pageY;
+      startTime = Date.now();
+    },
+    false
+  );
+
+  element.addEventListener(
+    'touchmove',
+    function (e) {
+      // Prevent dragging viewport
+      e.preventDefault();
+    },
+    false
+  );
+
+  element.addEventListener(
+    'touchend',
+    function (e) {
+      const touch = e.changedTouches[0];
+      const deltaX = touch.pageX - startX;
+      const deltaY = touch.pageY - startY;
+      const elapsedTime = Date.now() - startTime;
+
+      // Check that elapsed time is within specified, horizontal dist traveled >= threshold, and vertical dist traveled <= 100
+      const isSwipe =
+        elapsedTime <= allowedTime &&
+        Math.abs(deltaX) >= xMinThreshold &&
+        Math.abs(deltaY) <= yMaxThreshold;
+      if (isSwipe) {
+        const dir: HorizontalSwipeDirection = deltaX < 0 ? 'right' : 'left';
+        handler(dir);
+      }
+    },
+    false
+  );
+}
+
+function nextDictionaryTab(
+  currentDict: MajorDataSeries,
+  enabledTabs: Record<MajorDataSeries, boolean>
+): MajorDataSeries {
+  let i = allMajorDataSeries.indexOf(currentDict);
```

Comment:
I'm using the `allMajorDataSeries` as the authoritative ordering here, not sure the advisability of that. In `tabs.ts` the tab list is manually enumerated like so:

```ts
const sections: Array<{
    series: MajorDataSeries;
    renderIcon: () => SVGElement;
}> = [
    { series: 'words', renderIcon: renderBook },
    { series: 'kanji', renderIcon: renderKanjiIcon },
    { series: 'names', renderIcon: renderPerson },
];
```

Worth changing that to a map over `allMajorDataSeries` to ensure that the two never get out of sync? Or maybe exporting the `sections` var from `tabs.ts` and relying on it here rather than `allMajorDataSeries`?

Reply:
As you've noticed, when we click a tab at the moment, it calls a `onSwitchDictionary` callback which calls `showDictionary` in `ContentHandler`. That `showDictionary` function actually takes a `'next'` value and works out which dictionary to show.

I think we could make `showDictionary` also take a `prev` value and then allow `onSwitchDictionary` to pass `'next'` and `'prev'`.

What do you think?

Reply:
Oh, nice! Yeah, I think that will be a cleaner solution.

Reply:
Great, thank you!